quinn-proto/src/token.rs

```diff
@@ -43,46 +51,44 @@ impl RetryToken {
                 .unwrap_or(0),
         );
 
-        let signature_pos = buf.len();
         match address.ip() {
             IpAddr::V4(x) => buf.put_slice(&x.octets()),
             IpAddr::V6(x) => buf.put_slice(&x.octets()),
         }
         buf.write(address.port());
         retry_src_cid.encode_long(&mut buf);
 
-        let signature = key.sign(&buf);
-        // No reason to actually encode the IP in the token, since we always have the remote addr for an incoming packet.
-        buf.truncate(signature_pos);
-        buf.extend_from_slice(signature.as_ref());
-        buf
+        key.seal(&mut buf).unwrap();
+
+        let mut token = Vec::new();
+        token.put_slice(self.random_bytes);
+        token.put_slice(&buf);
+        token
     }
 
     pub fn from_bytes(
-        key: &impl HmacKey,
-        address: &SocketAddr,
-        retry_src_cid: &ConnectionId,
-        data: &[u8],
+        key: &impl AeadKey,
+        _address: &SocketAddr,
+        _retry_src_cid: &ConnectionId,
+        raw_token_bytes: &'a [u8],
     ) -> Result<Self, ()> {
+        if raw_token_bytes.len() < Self::RANDOM_BYTES_LEN {
+            // Invalid length
+            return Err(());
+        }
+
+        let random_bytes = &raw_token_bytes[..Self::RANDOM_BYTES_LEN];
+        let mut sealed_token = raw_token_bytes[Self::RANDOM_BYTES_LEN..].to_vec();
+        let data = key.open(&mut sealed_token)?;
         let mut reader = io::Cursor::new(data);
 
         let orig_dst_cid = ConnectionId::decode_long(&mut reader).ok_or(())?;
         let issued = UNIX_EPOCH + Duration::new(reader.get::<u64>().map_err(|_| ())?, 0);
 
-        let signature_start = reader.position() as usize;
-        let mut buf = Vec::new();
-        buf.put_slice(&data[0..signature_start]);
-        match address.ip() {
-            IpAddr::V4(x) => buf.put_slice(&x.octets()),
-            IpAddr::V6(x) => buf.put_slice(&x.octets()),
-        }
-        buf.write(address.port());
-        retry_src_cid.encode_long(&mut buf);
-
-        key.verify(&buf, &data[signature_start..]).map_err(|_| ())?;
```

Comment:
We must retain the logic to verify source address and retry source CID to defend against spoofing. That should be done with AAD as described in the above comment.